### ISSUE TITLE
Feedback on `needless_continue` in Rust 1.86

### ISSUE THREAD

**Issue by ojeda** [rust-lang/rust-clippy#14536]
### Description

Starting with Rust 1.86.0, Clippy's `needless_continue` lint complains about [the last statement of a loop](https://github.com/rust-lang/rust-clippy/pull/13891), including cases like:

    while ... {
        match ... {
            ... if ... => {
                ...
                return ...;
            }
            _ => continue,
        }
    }

as well as nested `match`es in a loop.

One solution in the cases we hit in the Linux kernel is changing `continue` for `()`, but arguably using `continue` shows the intent better when it is alone in an arm like that.

In the kernel in particular, I am not sure we want to force people to try to find other ways to write the code either, in cases when that applies.

Moreover, the help text does not really apply in at least some of the new cases the lint has introduced, e.g. in this one cannot simply "drop" the expression:

    warning: this `continue` expression is redundant
      --> rust/macros/helpers.rs:85:18
       |
    85 |             _ => continue,
       |                  ^^^^^^^^
       |
       = help: consider dropping the `continue` expression
       = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_continue
       = note: requested on the command line with `-W clippy::needless-continue`

The examples in the documentation do not show a case like this, either, so the second "help" line does not help.

In addition, locally disabling the lint is not possible with `expect`, since the behavior differs across versions. Using `allow` would be possible, but, even then, an extra line just for this is a bit too much, especially if there are other ways to satisfy the lint.

We would like to keep parts of this lint enabled in the kernel, since it does catch some issues that seem good to lint about, but others seem a bit too much. So it is likely we will globally disable this one for the moment.

Could perhaps this lint be split into some of its subcases? Some seem useful, and if most projects agree with some of those, then it may be easier to enable by default at some point.

Thanks!

### Version

```text
rustc 1.86.0 (05f9846f8 2025-03-31)
binary: rustc
commit-hash: 05f9846f893b09a1be1fc8560e33fc3c815cfecb
commit-date: 2025-03-31
host: x86_64-unknown-linux-gnu
release: 1.86.0
LLVM version: 19.1.7
```

**Comment by asibahi** [comment]
I have come across something to this as a false positive. I have this (honestly slightly ridiculous) `loop match` construct:

```rust
let mut counter = 2;
let buf = loop {
    match AGENT.get(link.as_str()).call() {
        Ok(mut res) => break res.body_mut().read_to_vec()?,
        Err(ureq::Error::Io(err))
            if counter > 0 && err.kind() == std::io::ErrorKind::ConnectionReset =>
        {   // probably not a good idea
            std::thread::sleep(std::time::Duration::from_millis(200));
            counter -= 1;
            continue; // <-- clippy pedantic: this continue expression is redundant
        },
        err => err?,
    };
};
```

but if I remove said `continue`, I get the following `rustc` error

```rust
error[E0308]: `match` arms have incompatible types
   --> mimiron/src/deck_image.rs:484:16
    |
475 | /     match AGENT.get(link.as_str()).call() {
476 | |         Ok(mut res) => break res.body_mut().read_to_vec()?,
477 | |         Err(ureq::Error::Io(err))
478 | |             if counter > 0 && err.kind() == std::io::ErrorKind::ConnectionReset =>
...   |
481 | |             counter -= 1;
    | |             ------------- this is found to be of type `()`
...   |
484 | |         err => err?,
    | |                ^^^^ expected `()`, found `Response<Body>`
485 | |     };
    | |_____- `match` arms have incompatible types
    |
    = note: expected unit type `()`
                  found struct `ureq::http::Response<ureq::Body>`

```

**Comment by samueltardieu** [comment]
@asibahi Could you open a specific issue for your case? This is a bug, distinct from @ojeda's feedback.
